### ISSUE TITLE
[Bug]: PostgreSQL buffer cache hit ratio averaging across DBs is misleading

### ISSUE THREAD

**Issue by shyamvalsan** [netdata/netdata#13506]
### Bug description

Buffer cache hit ratio is being averaged across all DBs present in Postgres server. 

This includes DBs which are idle and where no operations are taking place - these idle DBs have a cache hit ratio of 0% since nothing is going on. 

On averaging, these 0% DBs bring down the average and the overall cache hit ratio ends up being a misleading number.

### Expected behavior

Cache hit ratio should only be averaged across active DBs. If no data lookup itself is taking place then the 0% cache hit ratio of such a DB should not be included in the averaging.

### Steps to reproduce

1. Use PostgreSQL Go Collector to monitor a Postgres installation with multiple DBs.
2. Go to db buffer cache chart - see below screenshot, hit % is shown as 63% and miss as 1.2% which is misleading (Also doesn't add up to 100%)

![image](https://user-images.githubusercontent.com/24860547/183854950-6975bc56-852a-4d9c-be01-813195cd96af.png)

3. Group by database - here you can see that there are just 2 active databases, and the true average of buffer cache hit should be around 98% - but instead we're including the inactive DBs as well and ending up with 63%

![image](https://user-images.githubusercontent.com/24860547/183855511-0fb62cb8-8a49-404b-a711-02392ae51019.png)

### Installation method

kickstart.sh

### System info

```shell
Not applicable
```

### Netdata build info

```shell
Latest nightly
```

**Comment by shyamvalsan** [comment]
This is applicable for many metrics not just buffer cache hit ratio. 

Basically anything we're averaging across DBs and include inactive DBs will deflate the average. Also averaging percentage values across DBs in this case causes composite charts to show weird looking outputs where the sums do not add up to 100%.

![image](https://user-images.githubusercontent.com/24860547/184385071-0c659c07-a2d7-417d-8691-743452225067.png)

**Comment by shyamvalsan** [comment]
cc: @papazach @cakrit @sashwathn

**Comment by ktsaou** [comment]
We have to find a generic way to do this. Since this is required for many metrics, we need to find a rule at our metric types document for them.

**Comment by ktsaou** [comment]
Please don't implement this until we agree on a generic rule.

**Comment by shyamvalsan** [comment]
Discussion happened w/ @ktsaou @ralphm @amalkov @sashwathn @hugovalente-pm @shyamvalsan and some of the points discussed: 

- There needs to be a general rule on how to handle metrics where aggregation could lead to misleading values
- In the case of buffer cache, we could use (100 - miss%) to get the hit% but this will be a one-off custom logic and @ralphm suggested to wait until we agree on a general rule going forward. 
- The logic mentioned above of (100 - x%) will only work for a subset of metrics and will not work in many other cases.
- One option discussed was to make the value being collected "NULL" in the cases where it was invalid or lacking useful value - but @ktsaou pointed out that this is how the collector identifies the inability to collect values currently and we do not want to conflate the two.
- The possibility of needing another placeholder value to indicate "invalid" or "not applicable" which is distinct from "NULL" was discussed.
- There are certain conditions that are specific to the metric being collected which can define whether the metric value is meaningful or not. One option is to have a boolean flag per metric whose value is defined by this condition - by default flag is not set and metrics are collected as usual, if exclusion condition is triggered (For example: DB is inactive and no operations are taking place) then flag is set and metric collection doesn't happen (or happens with an "invalid" value).

**Comment by shyamvalsan** [comment]
@ktsaou @ralphm @ilyam8 and others please add your thoughts so we can take this discussion forward asynchronously.

**Comment by vkalintiris** [comment]
Driver-by comment: It sounds to me that this problem is similar to the one we've already solved in alarms, ie. specifying the lookup expression that defines the value of an alarm. I'd guess that we could follow a similar approach to allow users to override the default aggregation method of a chart for multiple hosts.

**Comment by ilyam8** [comment]
My thought is that "incorrect values" is questionable. Using this logic you can say the same for **every metric**. I think that averaging is perfectly fine (also, at what point does the DB become _active_?).

**Comment by ilyam8** [comment]
@shyamvalsan I don't understand why this is a bug in collectors and how I am supposed to fix it.

**Comment by shyamvalsan** [comment]
@ilyam8 there were discussions on changing the value collected to a NULL or something indicating Invalid/Not applicable - this is why it is in the agent/collectors area now. There could of course be other ways to fix this problem. 

The point regarding incorrect values is this. 

Currently if we look at our cache hit rate summary (here's a snapshot from Grafana using Netdata) it says 71% - in terms of databases and PostgreSQL specifically this is a really poor cache hit indicative of serious problems in our database. 

![image](https://user-images.githubusercontent.com/24860547/187630280-ff819a85-5764-45be-967f-2717fd050bc7.png)

While in reality our buffer rate is actually 100% or close to it. And there are just some databases we are monitoring where there is ZERO activity (hence zero hits and zero misses). 

![image](https://user-images.githubusercontent.com/24860547/187631027-09b811f2-42ad-4505-a894-ecea012c3c59.png)

IMO aggregating this to 71% is incorrect - and in any case it is misleading and not useful.

**Comment by ilyam8** [comment]
@shyamvalsan even if we consider this as a problem (which I don't see as a problem), that is applicable to a lot of metrics, not Postgres DB cache hit ratio specific.

> there were discussions on changing the value collected to a NULL or something indicating Invalid/Not applicable

Ok, but the fact the discussion happened doesn't mean I have to do something, right? I don't understand what "changing the value collected to a NULL" means - it is not currently supported by Netdata. If the idea was to skip collecting values (Netdata puts NULL in DB), it won't work too.

**Comment by shyamvalsan** [comment]
@ilyam8 

- Yes, this problem is not specific to Postgres DB cache hit ratio, it affects a lot of other metrics as well. And we should fix those too. The reason this metric was highlighted is because we cannot "launch" the PostgreSQL use-case with this value in its current state.

- Understand your concern, unassigning for the time being since we do not have an agreed way forward here. 

But @ktsaou @cakrit @ralphm @sashwathn we need a decision to be made soon or we'll start blocking this use-case as it will soon be ready.

**Comment by cakrit** [comment]
To begin with, we need a rule for the BE to implement for specific contexts, based on the value of a reserved label. So, no involvement from the collector side. Eventually that label will become a metric type.

**Comment by cakrit** [comment]
@papazach can you come up with a reasonable proposal to do this in the service that collects and aggregates the responses? Not sure who will implement it when, but I think you'll have the best idea on how to progress here.

**Comment by papazach** [comment]
To be honest I share the same opinion as @ilyam8 on this one:

>My thought is that "incorrect values" is questionable. Using this logic you can say the same for every metric. I think that averaging is perfectly fine (also, at what point does the DB become active?).

Additionally, the user knowing which DBs are _inactive_ and are rendering his composite view potentially _misleading_, could very well filter them out using the _filtered by_ **conditions** dropdown and selecting only the databases having "correct values". Am I missing something?

**Comment by shyamvalsan** [comment]
@papazach I have to disagree here. 

In this example, we have 4 dbs with only one of them active (with a cache hit ratio of 99%) but here's what Netdata shows on the composite chart. This makes no sense at all. I know that if you average the individual dbs (including the ones with hit ratio and miss ratio 0% because they were inactive) you will get this number, the problem is not in the value itself but in the fact that it is useless (and worse misleading).

![image](https://user-images.githubusercontent.com/24860547/188908161-9b96699d-2f77-409e-a4cc-153d90bc6a9c.png)

We have all the data on the cloud to make the right composite cache hit ratio, we just need to do disk/memory on the below composite information and we'd get the accurate composite cache hit ratio.

![image](https://user-images.githubusercontent.com/24860547/188908512-822579d2-2839-48f4-8da7-39250c58178b.png)

**Comment by shyamvalsan** [comment]
@papazach for further reference, here's a Grafana dashboard using Netdata metrics which shows an accurate composite cache hit ratio (this is from our production k8s room btw). Formula is: 
`composite cache hit ratio = ((sum(memory_reads) - sum(disk_reads))*100)/sum(memory_reads)`

https://grafana.corp.staging.netdata.cloud/goto/u4dzRoMVz?orgId=1

![image](https://user-images.githubusercontent.com/24860547/188913702-a3b9e476-9d28-402a-bdcc-70bd321d7145.png)


why shouldn't we be able to do this in netdata cloud as well, here's what Netdata cloud displays currently for the same k8s room in production

![image](https://user-images.githubusercontent.com/24860547/188913224-71c25338-66b1-49d0-a736-1747550046d3.png)

**Comment by hugovalente-pm** [comment]
@shyamvalsan considering the example you showed, what seems that would help solve this, and maybe other wrongly reported metrics, is allowing some of the flexibility of Grafana, making a composite chart out of metric that is combined/derived as a result of computations between different metrics - these formulas we would need to provide out of the box for the charts on Overview / Single Node view and maybe allow users to define custom charts combining different metrics on custom dashboards

**Comment by shyamvalsan** [comment]
yeah that's probably the right way to solve this and other similar issues. We should be able to create derived metrics on the cloud, without requiring any collector level changes.

@hugovalente-pm should we move this ticket to cloud repo?

**Comment by hugovalente-pm** [comment]
@shyamvalsan if others agree I think so, but would be good to also get @cakrit @papazach (who have recently been involved in the discussion)

**Comment by shyamvalsan** [comment]
Completely agree that providing workarounds/exceptions is not the right way forward - for the reasons you listed. Ideally this shouldn't be solved in a way specific to this particular metric (buffer cache hit ratio). 

Having derived metrics on the cloud (imo) offers a generic solution that goes beyond this particular metric.

**Comment by hugovalente-pm** [comment]
exactly, the idea of suggesting derived metrics on cloud is to avoid putting just some _nail_ to fix one specific metric. with such a feature we have the flexibility to _manipulate_ charts which have metrics that we consider aren't helpful or misleading as well as provide flexibility for users to create their own derived metrics on custom dashboards.

@papazach the goal is to not spec out a feature like this right now 😄 
we are trying to validate if we consider this a way to solve this issue and also a valuable feature that we think our product would benefit from it. 

if we are in agreement this will need to go through the prioritization process to understand, among other things, when we would tackle this.

**Comment by hugovalente-pm** [comment]
one additional point, we should evaluate if some of the calculations can be done on the agent before sending the data to Cloud, but probably we can only confirm this once we start thinking on some concrete examples

**Comment by sashwathn** [comment]
@hugovalente-pm : I think derived metrics is indeed the solution. In many cases we have 2 charts for essentially the same metric - one to visualise "raw values" and another to show the ratio / utilisation etc. By introducing derived metrics, we can make do with this and actually play around with aggregations etc which will be more meaningful.

**Comment by cakrit** [comment]
I don't understand how derived metrics can solve this generically, what exactly would we do?

@vkalintiris I didn't see an explanation of how you solved the problem with a lookup. How would it apply here?

An additional suggestion:
Add new parameter with a cloud UI option on each chart of the overview tab to "exclude zero values" and perhaps also "exclude negative values" or "exclude positive values" from the data returned from the agent. Allow for a context-level default setting for one of these parameters. So e.g. for the specific ratio we're discussing here, the context would indicate that we want to exclude zeroes. This parameter is sent to the BE in the request and the BE knows to follow the instruction.

**Comment by hugovalente-pm** [comment]
@cakrit the idea of derived metrics came from this comment from @shyamvalsan https://github.com/netdata/netdata/issues/13506#issuecomment-1239519315, where if a formula like this was applied we could solve the "wrong" value of the cache hit ratio

> composite cache hit ratio = ((sum(memory_reads) - sum(disk_reads))*100)/sum(memory_reads)


I'm not sure adding the chart options would help solve this issue since we may have values from the collector that could be a correct/valid zero and others that mean nothing was collected. passing the option to "exclude zero values" would exclude both

independent of the above, what you mention about chart options is an interesting feature to consider we had discussed these in the past when we thought about revisiting the Agent Dashboard and wanted to move some existing features to Cloud https://github.com/netdata/product/issues/2369

**Comment by cakrit** [comment]
The % metric that we were discussing before wasn't about cache hit %, it was returned rows vs rows fetched from the DB and would practically never have a 0% value in reality. 

I understand the concerns, but perfect is the enemy of good. If we can't think of anything that will let us produce valid aggregate charts without changing the values returned by the agent, it's time to extend the agent and the UI to support N/A (null) values. The % we saw in the postgreSQL use case was just wrong and misleading.
